### PR TITLE
WIP: Improve batching for document add / removes for reduced allocations

### PR DESCRIPTION
**Pull request by ToddGrun**
The modified entry points now only fork the solutionstate and project once when batched add/remove document calls are made.

Locally, I saw about a 3.5% allocation reduction when bringing up completion 20 times (half from ^j, half from typing a '.')

### CODE REVIEW

**Comment by ToddGrun** [comment]
profile images from local completion test:

before:
image

after:
image

File: src/Workspaces/Core/Portable/Workspace/Solution/SolutionCompilationState.cs [hunk 1]
```diff
@@ -1570,6 +1586,39 @@ public SolutionCompilationState RemoveAnalyzerConfigDocuments(ImmutableArray<Doc
         return newCompilationState;
     }
 
+    private static void AddToMultiDictionary<TKey, TValue, TData>(
+            ImmutableArray<TData> data,
+            Func<TData, TKey> keySelector,
+            Func<TData, TValue> valueSelector,
+            Dictionary<TKey, ArrayBuilder<TValue>> valuesByKey)
+            where TKey : notnull
```
Comment (CyrusNajmabadi) [inline src/Workspaces/Core/Portable/Workspace/Solution/SolutionCompilationState.cs:1594]: indentation.

File: src/Workspaces/Core/Portable/Workspace/Solution/SolutionCompilationState.cs [hunk 2]
```diff
@@ -1570,6 +1586,39 @@ public SolutionCompilationState RemoveAnalyzerConfigDocuments(ImmutableArray<Doc
...
+    {
+        // Collect the counts so we can preallocate the correct amount for each key.
+        using var _1 = PooledDictionary<TKey, int>.GetInstance(out var countsForKeys);
+        foreach (var d in data)
+        {
+            var key = keySelector(d);
+            if (!countsForKeys.TryGetValue(key, out var count))
+                countsForKeys[key] = 1;
+            else
+                countsForKeys[key] = count + 1;
```
Comment (CyrusNajmabadi) [inline src/Workspaces/Core/Portable/Workspace/Solution/SolutionCompilationState.cs:1604]:
```suggestion
            countsForKeys[key] = countsForKeys.TryGetValue(key, out var count) ? count + 1 : 1;
```

File: src/Workspaces/Core/Portable/Workspace/Solution/SolutionCompilationState.cs [hunk 3]
```diff
@@ -1473,18 +1473,25 @@ public SolutionCompilationState RemoveAnalyzerConfigDocuments(ImmutableArray<Doc
...
 
-        var documentInfosByProjectId = documentInfos.ToMultiDictionary(
-            static d => d.Id.ProjectId,
-            d =>
+        using var _1 = PooledDictionary<ProjectId, ArrayBuilder<TDocumentState>>.GetInstance(out var documentStatesByProjectId);
+
+        AddToMultiDictionary(
+            documentInfos,
+            keySelector: static documentInfo => documentInfo.Id.ProjectId,
+            valueSelector: documentInfo =>
```
Comment (CyrusNajmabadi) [inline src/Workspaces/Core/Portable/Workspace/Solution/SolutionCompilationState.cs:1481]: any desire to take a TArgs to allow this to be static?

File: src/Workspaces/Core/Portable/Workspace/Solution/SolutionCompilationState.cs [hunk 4]
```diff
@@ -1570,6 +1586,39 @@ public SolutionCompilationState RemoveAnalyzerConfigDocuments(ImmutableArray<Doc
...
+            if (!countsForKeys.TryGetValue(key, out var count))
+                countsForKeys[key] = 1;
+            else
+                countsForKeys[key] = count + 1;
+        }
+
+        foreach (var d in data)
+        {
+            var key = keySelector(d);
+            if (!valuesByKey.TryGetValue(key, out var valuesBuilder))
```
[the remark this answers is not in the document]
Comment (ToddGrun) [reply src/Workspaces/Core/Portable/Workspace/Solution/SolutionCompilationState.cs:1610]: I'm not clear on how this differs from what I'm doing. I don't think there would be any resizing with the current implementation.

[the remark this answers is not in the document]
Comment (CyrusNajmabadi) [reply src/Workspaces/Core/Portable/Workspace/Solution/SolutionCompilationState.cs:1610]: so, in the above code, if you don't find the value, you create a new one with teh right length to add all the values into.

however, say you *do* find the value.  Shouldn't you resize it once before adding all the items into it?

[the remark this answers is not in the document]
Comment (ToddGrun) [reply src/Workspaces/Core/Portable/Workspace/Solution/SolutionCompilationState.cs:1610]: Oh, I think I see the disconnect. You are worried about the case where the method is sent a valuesByKey collection that isn't empty. I'm not worried about that case as neither of the callers use the method in this way. If this method were moved to a more generic location, then I would be more inclined to defend against that usage.

**Pull request closed without merging**